tree/ntuple/v7/src/RPageStorageDaos.cxx

```diff
@@ -0,0 +1,462 @@
+   fClusterMinOffset = std::min(offsetData, fClusterMinOffset);
+   fClusterMaxOffset = std::max(offsetData + zippedBytes, fClusterMaxOffset);
+
+   if (!isAdoptedBuffer)
+      delete[] buffer;
+
+   RClusterDescriptor::RLocator result;
+   result.fPosition = offsetData;
+   result.fBytesOnStorage = zippedBytes;
+   return result;
+}
+
+
+ROOT::Experimental::RClusterDescriptor::RLocator
+ROOT::Experimental::Detail::RPageSinkDaos::CommitClusterImpl(ROOT::Experimental::NTupleSize_t /* nEntries */)
+{
+   RClusterDescriptor::RLocator result;
+   result.fPosition = fClusterMinOffset;
+   result.fBytesOnStorage = fClusterMaxOffset - fClusterMinOffset;
+   fClusterMinOffset = std::uint64_t(-1);
+   fClusterMaxOffset = 0;
+   return result;
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkDaos::CommitDatasetImpl()
+{
+   const auto &descriptor = fDescriptorBuilder.GetDescriptor();
+   auto szFooter = descriptor.SerializeFooter(nullptr);
+   auto buffer = std::unique_ptr<unsigned char []>(new unsigned char[szFooter]);
+   descriptor.SerializeFooter(buffer.get());
+
+   auto zipBuffer = std::unique_ptr<unsigned char[]>(new unsigned char[szFooter]);
+   auto szZipFooter = fCompressor(buffer.get(), szFooter, fOptions.GetCompression(),
+      [&zipBuffer](const void *b, size_t n, size_t o){ memcpy(zipBuffer.get() + o, b, n); } );
+   WriteNTupleFooter(zipBuffer.get(), szZipFooter, szFooter);
+   WriteNTupleAnchor();
+}
+
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleHeader(
+		const void *data, size_t nbytes, size_t lenHeader)
+{
+   fContainer->WriteObject(kOidHeader, data, nbytes, kDistributionKey, kAttributeKey);
+   fNTupleAnchor.fLenHeader = lenHeader;
+   fNTupleAnchor.fNBytesHeader = nbytes;
+   fNTupleAnchor.fSeekHeader = 0;
+}
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleFooter(
+		const void *data, size_t nbytes, size_t lenFooter)
+{
+   fContainer->WriteObject(kOidFooter, data, nbytes, kDistributionKey, kAttributeKey);
+   fNTupleAnchor.fLenFooter = lenFooter;
+   fNTupleAnchor.fNBytesFooter = nbytes;
+   fNTupleAnchor.fSeekFooter = 0;
+}
+
+void ROOT::Experimental::Detail::RPageSinkDaos::WriteNTupleAnchor() {
+   fContainer->WriteObject(kOidAnchor, &fNTupleAnchor, sizeof(fNTupleAnchor),
+                           kDistributionKey, kAttributeKey);
+}
+
+ROOT::Experimental::Detail::RPage
+ROOT::Experimental::Detail::RPageSinkDaos::ReservePage(ColumnHandle_t columnHandle, std::size_t nElements)
+{
+   if (nElements == 0)
+      nElements = kDefaultElementsPerPage;
+   auto elementSize = columnHandle.fColumn->GetElement()->GetSize();
+   return fPageAllocator->NewPage(columnHandle.fId, elementSize, nElements);
+}
+
+void ROOT::Experimental::Detail::RPageSinkDaos::ReleasePage(RPage &page)
+{
+   fPageAllocator->DeletePage(page);
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+
+
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageAllocatorDaos::NewPage(
```

Comment:
We probably don't need a custom allocator but we can simply use the heap page allocator.

Reply:
`RPageAllocatorHeap` does not allow to "adopt" a buffer (used in line 341).